execution/gethexec/espresso_finality_node.go

```diff
@@ -27,34 +30,113 @@ Espresso Finality Node creates blocks with finalized hotshot transactions
 type EspressoFinalityNode struct {
 	stopwaiter.StopWaiter
 
-	config     SequencerConfigFetcher
-	execEngine *ExecutionEngine
-	namespace  uint64
-
-	espressoClient  *espressoClient.Client
-	nextSeqBlockNum uint64
+	config                  SequencerConfigFetcher
+	namespace               uint64
+	executionEngine         *ExecutionEngine
+	espressoClient          *espressoClient.Client
+	nextSeqBlockNum         uint64
+	messagesWithMetadata    []*arbostypes.MessageWithMetadata
+	messagesWithMetadataPos []uint64
+	messagesStateMutex      sync.Mutex
 }
 
-func NewEspressoFinalityNode(execEngine *ExecutionEngine, configFetcher SequencerConfigFetcher) *EspressoFinalityNode {
+func NewEspressoFinalityNode(configFetcher SequencerConfigFetcher, execEngine *ExecutionEngine) *EspressoFinalityNode {
 	config := configFetcher()
 	if err := config.Validate(); err != nil {
-		panic(err)
+		log.Crit("Failed to validate espresso finality node config", "err", err)
 	}
 	return &EspressoFinalityNode{
-		execEngine:      execEngine,
 		config:          configFetcher,
 		namespace:       config.EspressoFinalityNodeConfig.Namespace,
 		espressoClient:  espressoClient.NewClient(config.EspressoFinalityNodeConfig.HotShotUrl),
 		nextSeqBlockNum: config.EspressoFinalityNodeConfig.StartBlock,
+		executionEngine: execEngine,
+	}
+}
+
+// TODO: For future versions, we should check the attestation quote to check if its from a valid TEE
+// TODO: Check for duplicates to make sure we don't double process the same message
+
+/**
+ * This function will create a block with the finalized hotshot transactions
+ */
+func (n *EspressoFinalityNode) createBlock() (returnValue bool) {
```

Comment:
What triggers this function to be called?

Reply:
Its an iterative function called from the `Start` function